jax/_src/tree_util.py

```diff
@@ -335,3 +335,126 @@ def __new__(klass, func, *args, **kw):
     lambda partial_: ((partial_.args, partial_.keywords), partial_.func),
     lambda func, xs: Partial(func, *xs[0], **xs[1]),  # type: ignore[index]
 )
+
+
+def broadcast_prefix(prefix_tree: Any, full_tree: Any,
+                     is_leaf: Optional[Callable[[Any], bool]] = None
+                     ) -> List[Any]:
+  # If prefix_tree is not a tree prefix of full_tree, this code can raise a
+  # ValueError; use prefix_errors to find disagreements and raise more precise
+  # error messages.
+  result = []
+  num_leaves = lambda t: tree_structure(t).num_leaves
+  add_leaves = lambda x, subtree: result.extend([x] * num_leaves(subtree))
+  tree_map(add_leaves, prefix_tree, full_tree, is_leaf=is_leaf)
+  return result
+
+def flatten_one_level(pytree: Any) -> Tuple[List[Any], Hashable]:
+  handler = _registry.get(type(pytree))
+  if handler:
+    children, meta = handler.to_iter(pytree)
+    return list(children), meta
+  elif isinstance(pytree, tuple) and hasattr(pytree, '_fields'):
+    return list(pytree), None
+  else:
+    raise ValueError(f"can't tree-flatten type: {type(pytree)}")
+
+def prefix_errors(prefix_tree: Any, full_tree: Any
+                  ) -> List[Callable[[str], ValueError]]:
+  return list(_prefix_error(KeyPath(()), prefix_tree, full_tree))
+
+class KeyPathEntry(NamedTuple):
+  key: Any
+  def pprint(self) -> str:
+    assert False  # must override
+
+class KeyPath(NamedTuple):
+  keys: Tuple[KeyPathEntry, ...]
+  def __add__(self, other):
+    if isinstance(other, KeyPathEntry):
+      return KeyPath(self.keys + (other,))
+    raise TypeError(type(other))
+  def pprint(self) -> str:
+    if not self.keys:
+      return ' tree root'
+    return ''.join(k.pprint() for k in self.keys)
+
+class GetitemKeyPathEntry(KeyPathEntry):
+  def pprint(self) -> str:
+    return f'[{self.key}]'
+
+class AttributeKeyPathEntry(KeyPathEntry):
+  def pprint(self) -> str:
+    return f'.{self.key}'
+
+class FlattenedKeyPathEntry(KeyPathEntry):  # fallback
+  def pprint(self) -> str:
+    return f'[<flat index {self.key}>]'
+
+def _child_keys(pytree: Any) -> List[KeyPathEntry]:
+  assert not treedef_is_leaf(tree_structure(pytree))
+  handler = _keypath_registry.get(type(pytree))
+  if handler:
+    return handler(pytree)
+  elif isinstance(pytree, tuple) and hasattr(pytree, '_fields'):
+    # handle namedtuple as a special case, based on heuristic
+    return [AttributeKeyPathEntry(s) for s in pytree._fields]
+  else:
+    num_children = len(treedef_children(tree_structure(pytree)))
+    return [FlattenedKeyPathEntry(i) for i in range(num_children)]
+
+_keypath_registry: Dict[Type, Callable[[Any], List[KeyPathEntry]]] = {}
```

Comment:
Is the plan that there will be other types registered than these? E.g., for custom pytrees?

Reply:
Yes, I hope so, at least for library-writers like Flax and Haiku (if they want to use this). Then those parameter data structures can get better error messages. WDYT?

Reply:
I am not sure I understand when would this be used? I can think of needing this for custom pytrees. It would be good to have a test/example.

Reply:
Well, libraries like Flax register lots of their own pytrees, e.g. for collections of parameters. Some examples are [`flax.struct`](https://flax.readthedocs.io/en/latest/flax.struct.html) and [`FrozenDict`](https://flax.readthedocs.io/en/latest/_modules/flax/core/frozen_dict.html#FrozenDict). In fact, that's the main motivating use case: several times users have gotten pytree errors when passing `PartitionSpec`s or similar for custom Flax pytrees, and the error messages are long and terrible!

From the point of view of this API, there's no difference between "custom pytrees" and pytrees which come registered with JAX. Even so, there are already tests which cover pytrees which don't come registered with JAX: see the use of `Special` in `test_fallback_keypath`. So I think the test coverage is sufficient.

I expect library developers like Flax folks will start using this API and add tests in their libraries for this functionality used with their own types.

Reply:
Googlers can check out go/bad-pytree-mismatch-error-message-example to see an example such message which came up recently.